### PR TITLE
Pass through strings using arrow2::Buffers

### PR DESCRIPTION
**Pull request by jleibs**
### What
This avoids copying strings and lets us end up with`&str` references pointed directly at the underlying arrow buffer.

1.8x speedup on mono_strings and 8.7x speedup on batch strings:
image

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/2931) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/2931)
- [Docs preview](https://rerun.io/preview/pr%3Ajleibs%2Fprimitive_buffers/docs)
- [Examples preview](https://rerun.io/preview/pr%3Ajleibs%2Fprimitive_buffers/examples)

### CODE REVIEW

File: crates/re_types/src/arrow_adapter.rs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+
+/// Convenience-wrapper around an arrow [`Buffer`] that is known to contain a
+/// string.
+///
+/// The arrow2 [`Buffer`] object is internally reference-counted and can be
+/// easily converted back to a `&str` referencing the underlying storage.
+/// This avoids some of the lifetime complexities that would otherwise
+/// arise from returning a `&str` directly, but is significantly more
+/// performant than doing the full allocation necessary to return a `String`.
+#[derive(Clone, Debug, Default)]
```
Comment (teh-cmc) [inline crates/re_types/src/arrow_adapter.rs:11]: `Hash` would be nice too

File: crates/re_types/src/arrow_adapter.rs
Comment (teh-cmc) [inline crates/re_types/src/arrow_adapter.rs:1]: `arrow_string.rs`?

File: crates/re_types/src/lib.rs [hunk 2]
```diff
@@ -502,5 +502,7 @@ mod size_bytes;
 pub use component_name::ComponentName;
 pub use size_bytes::SizeBytes;
 
+pub mod arrow_adapter;
```
Comment (teh-cmc) [inline crates/re_types/src/lib.rs:505]: Only expose `ArrowString` itself so the details of the module hierarchy don't become part of the public API.

```suggestion
mod arrow_adapter;
pub use self::arrow_adapter::ArrowString;

File: crates/re_types_builder/src/codegen/rust.rs [hunk 3]
```diff
@@ -1791,23 +1791,23 @@ fn quote_arrow_field_serializer(
                         quote! {
                             .flat_map(|datum| {
                                 let #quoted_binding = datum;
-                                #quoted_data_dst .bytes()
+                                #quoted_data_dst .0.clone()
```
Comment (teh-cmc) [inline crates/re_types_builder/src/codegen/rust.rs:1794]:
```suggestion

                                // NOTE: `Buffer::clone`, which is just a ref-count bump
                                #quoted_data_dst .0.clone()

File: crates/re_types_builder/src/codegen/rust.rs [hunk 4]
```diff
@@ -1791,23 +1791,23 @@ fn quote_arrow_field_serializer(
...
-                                #quoted_data_dst.len()
+                                #quoted_data_dst.0.len()
                             }).unwrap_or_default()
                         },
                     )
                 } else {
                     (
                         quote! {
-                            .flat_map(|s| s.bytes())
+                            .flat_map(|s| s.0.clone())
```
Comment (teh-cmc) [inline crates/re_types_builder/src/codegen/rust.rs:1807]:
```suggestion
                            // NOTE: `Buffer::clone`, which is just a ref-count bump
                            .flat_map(|s| s.0.clone())

File: crates/re_types_builder/src/codegen/rust.rs [hunk 5]
```diff
@@ -2307,21 +2307,25 @@ fn quote_arrow_field_deserializer(
...
-                    .downcast_ref::<Utf8Array<i32>>()
-                    .unwrap() // safe
-                    .into_iter()
+                {
+                let downcast = #data_src.as_any().downcast_ref::<Utf8Array<i32>>().unwrap();
+                let offsets = downcast.offsets();
+                arrow2::bitmap::utils::ZipValidity::new_with_validity(
+                    offsets.iter().zip(offsets.lengths()),
+                    downcast.validity(),
+                )
```
Comment (teh-cmc) [inline crates/re_types_builder/src/codegen/rust.rs:2325]: Nice, I probably could've used that in a few places :sweat_smile:

**Review by teh-cmc — APPROVED** [review]
nice!

**Comment by emilk** [comment]
Does this close https://github.com/rerun-io/rerun/issues/1887 ?

**Comment by jleibs** [comment]
> Does this close https://github.com/rerun-io/rerun/issues/1887 ?

It's at least related. This only affects code-generated strings, so we need to migrate TextEntry to codegen. Also need to take a look at how the TextEntry query results are used to make sure we aren't gratuitously converting them back to strings somewhere in the query path.

**Pull request merged**